src/ethereum/homestead/vm/interpreter.py

```diff
@@ -249,3 +249,80 @@ def execute_code(message: Message, env: Environment) -> Evm:
         evm.has_erred = True
     finally:
         return evm
+
+
+def collect_accounts_to_delete(evm: Evm) -> Set[Address]:
+    """
+    Collects all the accounts that were marked for deletion by the
+    `SELFDESTRUCT` opcode.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Returns
+    -------
+    accounts_to_delete: `set`
+        returns all the accounts need marked for deletion by the
+        `SELFDESTRUCT` opcode.
+    """
+    if evm.has_erred:
+        return set()
+    else:
+        return set(
+            chain(
+                evm.accounts_to_delete,
+                *(collect_accounts_to_delete(child) for child in evm.children),
+            )
+        )
+
+
+def collect_logs(evm: Evm) -> Tuple[Log, ...]:
+    """
+    Collects all the logs emitted while executing the message call.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Returns
+    -------
+    logs: `tuple`
+        returns all the logs that were emitted while executing the
+        message call.
+    """
+    if evm.has_erred:
+        return ()
+    else:
+        return tuple(
+            chain(
+                evm.logs,
+                *(collect_logs(child_evm) for child_evm in evm.children),
+            )
+        )
```

Comment:
Similarly, maybe something like:

```python
return evm.logs + tuple(collect_logs(child_evm) for child_evm in evm.children)
```

Reply:
I think I tried this and it didn't work. I used the chain method to flatten the tuple, otherwise, this code generates nested tuples.

Reply:
Ah, right, it's nested tuples. My bad!